validation/standard/standard-versions.toml

```diff
@@ -1,14 +1,38 @@
-["op-contracts/v1.4.0"] # Fault Proofs https://github.com/ethereum-optimism/optimism/releases/tag/op-contracts%2Fv1.4.0
-address_manager = ""
-l1_cross_domain_messenger = "2.3.0"
-l1_erc721_bridge = "2.1.0"
-l1_standard_bridge = "2.1.0"
-optimism_mintable_erc20_factory = "1.9.0"
-optimism_portal = "3.10.0"
-system_config = "2.2.0"
-protocol_versions = "1.0.0"
-anchor_state_registry = "1.0.0"
-delayed_weth = "1.0.0"
-dispute_game_factory = "1.0.0"
-fault_dispute_game = "1.2.0"
-permissioned_dispute_game = "1.2.0"
+op_contracts_release = "op-contracts/v1.4.0"
+
+[contract_tag]
+
+# Fault Proofs https://github.com/ethereum-optimism/optimism/releases/tag/op-contracts%2Fv1.4.0
+[contract_tag."op-contracts/v1.4.0"]
+address_manager = { version = "" }
+l1_cross_domain_messenger = { version = "2.3.0" }
+l1_erc721_bridge = { version = "2.1.0" }
+l1_standard_bridge = { version = "2.1.0" }
+optimism_mintable_erc20_factory = { version = "1.9.0" }
+optimism_portal = { version = "3.10.0" }
+system_config = { version = "2.2.0" }
+protocol_versions = { version = "1.0.0" }
+anchor_state_registry = { version = "1.0.0" }
+delayed_weth = { version = "1.0.0" }
+dispute_game_factory = { version = "1.0.0" }
+fault_dispute_game = { version = "1.2.0" }
+permissioned_dispute_game = { version = "1.2.0" }
+mips = { version = "1.0.1" }
+preimage_oracle = { version = "1.0.0" }
```

Comment:
For FP contracts:
- OptimismPortal2, DisputeGameFactory, and DelayedWETH have shared implementations for all chains, so should have an address here
- PreimageOracle and MIPS are singletons (one, non-proxied instance for all chains). Therefore these probably should be at the superchain level with SuperchainConfig and ProtocolVersions
- Anchor state registry, fault dispute game permissioned dispute game: we know what semver to expect, but each one has a unique address per chain (i.e. these contracts cannot be reused)

Reply:
> Anchor state registry, fault dispute game permissioned dispute game: we know what semver to expect, but each one has a unique address per chain (i.e. these contracts cannot be reused)

Retained only semvers, dropped addresses.

> PreimageOracle and MIPS are singletons (one, non-proxied instance for all chains). Therefore these probably should be at the superchain level with SuperchainConfig and ProtocolVersions

@mds1 It appears that the contract addresses for MIPS and PreimageOracle are different for `v1.4.0` and `v1.6.0`. As is, there is no way to disambiguate contract addresses at the superchain level for different releases. Should we keep these addresses in `standard-versions.toml` so we have the ability to store an address per release?

Reply:
> Should we keep these addresses in `standard-versions.toml` so we have the ability to store an address per release?

Yea this seems like a good solution. I would just suggest that we have some way to differentiate the implementations from the singletons, e.g. maybe in the toml we use `[contract_tag."op-contracts/v1.4.0".implementations]` and `[contract_tag."op-contracts/v1.4.0".singletons]`, or you can add a third key like `{ version ="foo", address = "bar", kind = <"implementation"|"singleton">

Reply:
Here's how we ended up codifying this https://github.com/ethereum-optimism/superchain-registry/pull/587